src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTag.java

```diff
@@ -12,7 +12,7 @@
 
 /**
  * Testing tag writing
- * @author Daniel Grenner // violation
+ * @author must match pattern 'S'
```

Comment:
At the top of the page, in the multiline comment, it is given as ```tagFormat = \S``` .

1. What does it mean.
2. Why is it present there? OR why do we need it?

Presence of this format is forcing me to remove ``` Daniel Grenner // violation``` and write ``` must match patter 'S'```. 

What is funny is that it removes the test error and the test case passes without problem. 

Other very important question:

1. In ```must match pattern 'S'```, what pattern are we talking about? What kind of pattern does ```'S'``` indicate.

I need to know this all to solve the issue. 
Also I'm quite sure that though what I have done removed the test error but totally isin't right as the violation message is absent and the message written in place is totally inconsistent with message written in ```javadoc\messages.properties``` which is ```javadoc.writeTag=Javadoc tag {0}={1}```.

Kindly help.

Reply:
it means that there should be atleast one not space
https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html search for "\S" in this page.

Reply:
make a line `* @author Daniel Grenner // violation 'Javadoc tag .*'`  ... it should work

Reply:
or better to make:
```
// violation 2 line below 'Javadoc tag .*'
/**
 * Testing tag writing
 * @author Daniel Grenner

```

Reply:
I have tried mulitple way to remove the error and have separately studied the errors. I've the got the following:

1. Not changing anything in the input file gives me the following error

```
/**
 * Testing tag writing
 * @author Daniel Grenner // violation
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
```

Error:
```
expected: …utWriteTag.java:15: Javadoc tag @author=Daniel Grenner // violation
but was : …utWriteTag.java:15: Type Javadoc tag @author must match pattern 'S'.
```

2. Updating the error message according how @romani has told, I get the following:

```
/**
 * Testing tag writing
 * @author Daniel Grenner // violation 'Javadoc tag .*'
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
```
Error:
```
expected: …utWriteTag.java:15: Javadoc tag @author=Daniel Grenner // violation
but was : …utWriteTag.java:15: Type Javadoc tag @author must match pattern 'S'.
```

3. Updating it again as it has been told:

```
package com.puppycrawl.tools.checkstyle.checks.javadoc.writetag;
// violation 2 line below 'Javadoc tag .*'
/**
 * Testing tag writing
 * @author Daniel Grenner
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
```
Error:
```
Violation lines for C:\Users\shubh\OneDrive\Desktop\checkstyle project\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\javadoc\writetag\InputWriteTag.java differ.
expected to be empty
but was: [15]
```
4. Making some changes on my own to remove the error I get:
```
package com.puppycrawl.tools.checkstyle.checks.javadoc.writetag;
// violation 3 lines below 'Javadoc tag .*'
/**
 * Testing tag writing
 * @author Daniel Grenner
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
```
Error:
```
expected: …utWriteTag.java:15: Javadoc tag @author=Daniel Grenner // violation
but was : …utWriteTag.java:15: Type Javadoc tag @author must match pattern 'S'.
```
5. Changing the violation message a little bit:
```
// violation 3 lines below 'Javadoc tag '@author'='Daniel Grenner''
/**
 * Testing tag writing
 * @author Daniel Grenner
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
```
Error:
```
Actual and expected violations differ.
expected to match: 15:(?:\d+:)?\s.*Javadoc tag '@author'='Daniel Grenner'.*
but was          : 15: Type Javadoc tag @author must match pattern 'S'.
```

So I am not able to understand the error ```Type Javadoc tag @author must match pattern 'S'```. 
@romani Kindly assist.

Reply:
If I am able to solve this error. I will most probably solve all errors

Reply:
Please make CLI execution on such code to see real violation message.

https://checkstyle.org/config_javadoc.html#WriteTag_Violation_Messages

I see multiple violation messages are used by Check. There might be 2 violations on same line.

Reply:
```
PS C:\Users\shubh\OneDrive\Desktop\checkstyle project> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
       "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
        <property name="charset" value="UTF-8" />
      <module name="TreeWalker">
        <module name="WriteTag">
            <property name="tag" value="@author"/>
               <property name="tokens"
            value="INTERFACE_DEF, CLASS_DEF, ENUM_DEF, ANNOTATION_DEF, RECORD_DEF" />
        <property name="tagFormat" value="\S"/>
  <property name="tagSeverity" value="info"/>
        </module>
      </module>
</module>


PS C:\Users\shubh\OneDrive\Desktop\checkstyle project> cat test.java
package com.puppycrawl.tools.checkstyle.checks.javadoc.writetag;

/**
 * Testing tag writing
 * @author Daniel Grenner // violation
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
class InputWriteTag
{
    /**
     * @todo Add a constructor comment
     */
    public InputWriteTag()
    {
    }

    public void method()
    {
    }

    /**
     * @todo Add a comment
     */
    public void anotherMethod()
    {
    }
}
PS C:\Users\shubh\OneDrive\Desktop\checkstyle project> java -jar "C:\Users\shubh\Downloads\checkstyle-10.3.4-all.jar" -c config.xml test.java
Starting audit...
[INFO] C:\Users\shubh\OneDrive\Desktop\checkstyle project\test.java:5: Javadoc tag @author=Daniel Grenner // violation [WriteTag]
Audit done.

```

This what I am getting.

Reply:
@romani note that the code and the errors that I showed previously for the violation line(https://github.com/checkstyle/checkstyle/pull/12288#discussion_r998914707) was actually the error that got displayed in my intellij IDEA when I ran the tests individually through intellij. I didn't separately run the config and get those errors until you separately told me to do so in the previous comment of yours(https://github.com/checkstyle/checkstyle/pull/12288#discussion_r999066617).

Reply:
![image](https://user-images.githubusercontent.com/90137881/196669855-09ef4c87-95d4-4697-9761-6bf4e7d76fbe.png)


It keeps giving me this error. What am I supposed to do to make it ```match pattern 'S'```

Reply:
Please put `// violation` above comment block and run CLI one more time.
You receive completely different error message from Check.
Please run Junit test in debug to understand why it happens.

Reply:
```
PS C:\Users\shubh\OneDrive\Desktop\checkstyle project> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
       "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
        <property name="charset" value="UTF-8" />
      <module name="TreeWalker">
        <module name="WriteTag">
            <property name="tag" value="@author"/>
               <property name="tokens"
            value="INTERFACE_DEF, CLASS_DEF, ENUM_DEF, ANNOTATION_DEF, RECORD_DEF" />
               <property name="tagFormat" value="\S"/>
          <property name="tagSeverity" value="info"/>
        </module>
      </module>
</module>


PS C:\Users\shubh\OneDrive\Desktop\checkstyle project> cat test.java
/*
WriteTag
tag = @author
tagFormat = \S
tagSeverity = (default)info
tokens = (default)INTERFACE_DEF, CLASS_DEF, ENUM_DEF, ANNOTATION_DEF, RECORD_DEF


*/
package com.puppycrawl.tools.checkstyle.checks.javadoc.writetag;
// violation 3 lines below 'Javadoc tag .*'
/**
 * Testing tag writing
 * @author Daniel Grenner
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
class InputWriteTag
{
    /**
     * @todo Add a constructor comment
     */
    public InputWriteTag()
    {
    }

    public void method()
    {
    }

    /**
     * @todo Add a comment
     */
    public void anotherMethod()
    {
    }
}
PS C:\Users\shubh\OneDrive\Desktop\checkstyle project> java -jar "C:\Users\shubh\Downloads\checkstyle-10.3.4-all.jar" -c config.xml test.java
Starting audit...
[INFO] C:\Users\shubh\OneDrive\Desktop\checkstyle project\test.java:14: Javadoc tag @author=Daniel Grenner [WriteTag]
Audit done.

```

Reply:
Ok, very strange but it only means we missing some nuance and execution from test and cli have different input file or config.

Please debug failing Junit to find why we have different behavior.

Reply:
to fix your problem apply patch:
```
$ git diff
diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java
index a51863c..482a70b 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/WriteTagCheckTest.java
@@ -58,7 +58,7 @@ public class WriteTagCheckTest extends AbstractModuleTestSupport {
     @Test
     public void testTag() throws Exception {
         final String[] expected = {
-            "15: " + getCheckMessage(MSG_WRITE_TAG, "@author", "must match pattern 'S'"),
+            "15: " + getCheckMessage(MSG_WRITE_TAG, "@author", "Daniel Grenner"),
         };
         verifyWithInlineConfigParser(getPath("InputWriteTag.java"), expected);
     }
diff --git a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTag.java b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTag.java
index b1e2fd3..bcb5747 100644
--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTag.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/writetag/InputWriteTag.java
@@ -1,7 +1,7 @@
 /*
 WriteTag
 tag = @author
-tagFormat = \S
+tagFormat = \\S
 tagSeverity = (default)info
 tokens = (default)INTERFACE_DEF, CLASS_DEF, ENUM_DEF, ANNOTATION_DEF, RECORD_DEF
 
@@ -12,7 +12,7 @@ package com.puppycrawl.tools.checkstyle.checks.javadoc.writetag;
 
 /**
  * Testing tag writing
- * @author must match pattern 'S'
+ * @author Daniel Grenner
  * @incomplete This class needs more code...
  * @doubletag first text
  * @doubletag second text

```

Reply:
@romani could you please confirm as to how to apply the patch. It is slighly unclear from the cli you have shared as to what is supposed to be entered.

Reply:
Make line as `tagFormat = \\S`

Reply:
Try to google how to apply patch to see impressive videos and manuals.

Reply:
@romani I was able to remove some of the error from the test cases (I ran the tests through intellij idea). But I'm still stuck on some of the test cases.

One very strange error I got was in the following test:

```
@Test
    public void testMethod() throws Exception {
        final String[] expected = {
            "24: " + getCheckMessage(MSG_WRITE_TAG, "@todo",
                    "Add a constructor comment"),
            "35: " + getCheckMessage(MSG_WRITE_TAG, "@todo", "Add a comment"),
        };
        verifyWithInlineConfigParser(getPath("InputWriteTagMethod.java"), expected);
    }
```

The input file is;

```
/*
WriteTag
tag = @todo
tagFormat = \\S
tagSeverity = (default)info
tokens = INTERFACE_DEF, CLASS_DEF, METHOD_DEF, CTOR_DEF


*/

package com.puppycrawl.tools.checkstyle.checks.javadoc.writetag;

/**
 * Testing tag writing
 * @author Daniel Grenner
 * @incomplete This class needs more code...
 * @doubletag first text
 * @doubletag second text
 * @emptytag
 */
class InputWriteTagMethod
{
    // violation 2 lines below , 'Javadoc tag .*'
    /**
     * @todo Add a constructor comment
     */
    public InputWriteTagMethod()
    {
    }

    public void method() // ok
    {
    }

    // violation 2 lines below , 'Javadoc tag .*'
    /**
     * @todo Add a comment
     */
    public void anotherMethod()
    {
    }
}
```

The error is :

```
Violation lines for C:\Users\shubh\OneDrive\Desktop\checkstyle project\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\javadoc\writetag\InputWriteTagMethod.java differ.
expected to be empty
but was: [21, 25]

```

Here line 21 is ```class InputWriteTagMethod```. I don't understand why is it reporting violation in that line evethough that line has a commit of ```//ok```

Reply:
@romani please checkout these issues

Reply:
Please push code to GitHub to let me see exact code and exact failure message

Reply:
@romani it has been pushed already

Reply:
@romani changing tagformat from \S to \\S helped solve some of the test cases. Also writing a proper violation message removed some errors as well. But don't know how to remove errors for other test cases.

Reply:
Push latest code to GitHub to let me see all and reproduce issue on my local.

Reply:
@romani Done.